lib/Partitioner/Partitioner.cpp

```diff
@@ -139,6 +139,147 @@ static BFSLevel getBFSLevel(Function *F) {
+        flag = false;
+        break;
+      }
+    }
+    if (flag) {
+      // This partition only has one successor.
+      Function *cur = (*it).first;
+      Function *suc = partitions[outUsers[0]];
+      NodesSet tmp = nodesSet.lookup(suc);
+      GraphMemInfo cost1 = partitions.getCost(cur);
+      GraphMemInfo cost2 = partitions.getCost(suc);
+      if (cost1.constMem + cost1.inMem + cost2.constMem + cost2.inMem -
+              cost1.outMem <
+          availableMemory) {
+        // We can combine the two partitions to fit one device.
+        for (NodesSet::iterator it2 = tmp.begin(); it2 != tmp.end(); ++it2) {
+          partitions.add(*it2, cur);
+        }
+        (*it).second.insert(tmp.begin(), tmp.end());
+        partitions.deletePartition(suc);
+        nodesSet.erase(suc);
+        module_->eraseFunction(suc);
+      }
+    }
+  }
+}
+
+void Partitioner::partitionsAdjust(NodeToFunctionMap &partitions,
+                                   uint64_t availableMemory) {
+  // For each partitioin, create a node set.
+  PartitionNodesMap nodesSet;
+  for (Map::iterator it = partitions.begin(); it != partitions.end(); ++it) {
+    nodesSet[(*it).second].insert((*it).first);
+  }
+
+  // Initial the memory cost for each partition. Now we use the output size to
+  // represent the communication cost.
+  for (PartitionNodesMap::iterator it = nodesSet.begin(); it != nodesSet.end();
+       ++it) {
+    GraphMemInfo cost = getGraphMemInfo((*it).second);
+    partitions.setCost((*it).first, cost);
+  }
+
+  // Move/Exchange nodes between any two connected partions, until no gain is
+  // get.
+  // Step1 Move: Assume Partition1 -> Partition2, try to move nodes from
+  // Partition2 to Partition1 if those nodes only use the nodes in
+  // Partition1(recursively) and the move won't make Partition1's memory exceeds
+  // the memory constratint, and the communication cost is minimized.
+  bool gain = true;
+  while (gain) {
+    // gain is initialized as false, it will be set to be true if there is at
+    // least one node can be moved from one set to antoher set.
+    gain = false;
+    for (PartitionNodesMap::iterator it = nodesSet.begin();
+         it != nodesSet.end(); ++it) {
+      NodesSet nSet = (*it).second;
+      std::vector<Node *> outUsers = getOutUsersWithOnePre(nSet);
+      if (outUsers.empty()) {
+        continue;
+      }
+      Function *cur = (*it).first;
+      uint64_t mem =
+          partitions.getCost(cur).constMem + partitions.getCost(cur).inMem;
+      uint64_t communicationCost = partitions.getCost(cur).outMem;
+      // Check if a node can be moved to current node set (i.e nSet).
+      for (int i = 0, e = outUsers.size(); i < e; i++) {
+        // Rule 1: this move won't break memory constraint.
+        if (memUsage_[outUsers[i]] + mem > availableMemory) {
+          continue;
+        }
+        // Rule 2: this move won't cause constant dupication.
```

Comment:
Question: May be it is OK to have constants duplication if the size of a constant is below a certain threshold?

Reply:
> Question: May be it is OK to have constants duplication if the size of a constant is below a certain threshold?

@opti-mix Theoretically, it should be allowed if it reduce the communication cost a lot or good for workload balance.  Will think about it for the next step. Thanks!